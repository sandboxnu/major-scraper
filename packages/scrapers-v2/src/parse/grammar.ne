@{%
// define custom tokens
// basically just check against the HRowType enum value
const HEADER = { test: x => x.type === "HEADER" };
const SUBHEADER = { test: x => x.type === "SUBHEADER" };
const COMMENT = { test: x => x.type === "COMMENT" };

const OR_COURSE = { test: x => x.type === "OR_COURSE" };
const OR_OF_AND_COURSE = { test: x => x.type === "OR_OF_AND_COURSE" };

const AND_COURSE = { test: x => x.type === "AND_COURSE" };

const PLAIN_COURSE = { test: x => x.type === "PLAIN_COURSE" };
const RANGE_LOWER_BOUNDED = { test: x => x.type === "RANGE_LOWER_BOUNDED" };
const RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS = { test: x => x.type === "RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS" };
const RANGE_BOUNDED = { test: x => x.type === "RANGE_BOUNDED" };
const RANGE_BOUNDED_WITH_EXCEPTIONS = { test: x => x.type === "RANGE_BOUNDED_WITH_EXCEPTIONS" };
const RANGE_UNBOUNDED = { test: x => x.type === "RANGE_UNBOUNDED" };
%}

@{%
// import postprocessors
const postprocess = require("./postprocess");
%}

# main entrypoint
main -> requirement2 :+                               {% id %}
requirement2 ->
    orCourse {% id %}
  | andCourse {% id %}
  | course {% id %}
  | range {% id %}

course -> %PLAIN_COURSE                               {% id %}
range ->
    %RANGE_LOWER_BOUNDED {% id %}
  | %RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS {% id %}
  | %RANGE_BOUNDED {% id %}
  | %RANGE_BOUNDED_WITH_EXCEPTIONS {% id %}
  | %RANGE_UNBOUNDED                                  {% id %}

# an OR always follows a plainCourse or andCourse
orCourse ->
  (course {% id %} | andCourse {% id %}) (%OR_COURSE {% id %} | %OR_OF_AND_COURSE {% id %}) :+
andCourse -> %AND_COURSE :+