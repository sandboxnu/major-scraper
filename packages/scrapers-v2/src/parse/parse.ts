import { CompiledRules, Grammar, Parser } from "nearley";

// at runtime, generate the ./grammar.ts file from the grammar.ne file
// eslint-disable-next-line @typescript-eslint/no-var-requires
const grammar: CompiledRules = require("./grammar");
// declare const grammar: CompiledRules;

const parser = new Parser(Grammar.fromCompiled(grammar));

parser.feed("");
// parser.results is an array of possible results
