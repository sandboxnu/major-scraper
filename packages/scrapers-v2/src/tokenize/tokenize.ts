import { assertUnreachable } from "@graduate/common";
import {
  appendPath,
  ensureLength,
  ensureLengthAtLeast,
  loadHTML,
  parseText,
} from "../utils";
import {
  COURSE_REGEX,
  RANGE_BOUNDED,
  RANGE_LOWER_BOUNDED_MAYBE_EXCEPTIONS_1,
  RANGE_LOWER_BOUNDED_MAYBE_EXCEPTIONS_2,
  RANGE_LOWER_BOUNDED_PARSE,
  RANGE_UNBOUNDED,
  SUBJECT_REGEX,
} from "./constants";
import {
  CourseRow,
  HDocument,
  HRow,
  HRowType,
  HSection,
  MultiCourseRow,
  RangeBoundedRow,
  RangeLowerBoundedRow,
  RangeUnboundedRow,
  TextRow,
  WithExceptions,
} from "./types";

/**
 * Fetch html for page and convert into intermediate representation (IR)
 * @param url the url of the page to tokenize
 */
export const fetchAndTokenizeHTML = async (url: string): Promise<HDocument> => {
  return tokenizeHTML(await loadHTML(url));
};

/**
 * Tokenize scraped html into intermediate representation (IR)
 * @param $ the cheerio static for the page to tokenize
 */
export const tokenizeHTML = async ($: CheerioStatic): Promise<HDocument> => {
  const majorName: string = parseText($("#site-title").find("h1"));
  const catalogYear: string = parseText($("#edition")).split(" ")[0];
  const yearVersion: number = parseInt(catalogYear.split("-")[0]);

  const requirementsContainer = $("#programrequirementstextcontainer");
  const sections = tokenizeSections($, requirementsContainer);

  const programRequiredHeading = requirementsContainer
    .find("h2")
    .filter(
      (_idx: number, element: CheerioElement) =>
        parseText($(element)).includes("Program") &&
        parseText($(element)).includes("Requirement")
    );
  const programRequiredHours =
    Number(parseText(programRequiredHeading.next()).split(/[\s\xa0]+/)[0]) || 0;

  return {
    programRequiredHours,
    yearVersion,
    majorName,
    sections: await sections,
  };
};

/**
 * Produces overall HSections for each HTML table in the page
 * @param $
 * @param requirementsContainer
 */
const tokenizeSections = async (
  $: CheerioStatic,
  requirementsContainer: Cheerio
): Promise<HSection[]> => {
  // use a stack to keep track of the course list title and description
  const descriptions: string[] = [];
  const courseList: HSection[] = [];

  for (const element of requirementsContainer.children().toArray()) {
    if (element.name === "h2" || element.name === "h3") {
      // element is h2 or h3 means it's a header text
      descriptions.push(parseText($(element)));
    } else if (
      element.name === "table" &&
      element.attribs["class"] === "sc_courselist"
    ) {
      // class "sc_courselist" signifies that this table is a list of courses
      // => parse the table's rows
      const tableDesc = descriptions.pop() || "";
      const courseTable = {
        description: tableDesc,
        entries: tokenizeRows($, element),
      };
      courseList.push(courseTable);
    } else if (
      // only necessary for business concentrations
      element.name === "ul" &&
      parseText($(element).prev()).includes("concentration")
    ) {
      // if we encounter an unordered list and preceding element contains text "concentration",
      // assume the list is of links for business concentrations.
      // only applies to:
      // https://catalog.northeastern.edu/undergraduate/business/business-administration-bsba/#programrequirementstext
      const links = constructNestedLinks($, element);
      const pages = await Promise.all(links.map(loadHTML));
      const containerId = "#concentrationrequirementstextcontainer";
      const concentrations = await Promise.all(
        pages.map((concentrationPage) =>
          tokenizeSections(concentrationPage, concentrationPage(containerId))
        )
      );
      courseList.push(...concentrations.flat());
    }
  }

  return courseList;
};

/**
 * Finds and fetches nested links, for majors with concentration requirements
 * on separate pages.
 * @param $
 * @param element
 */
const constructNestedLinks = ($: CheerioStatic, element: CheerioElement) => {
  // TODO: add support to non-current catalogs
  const base = "https://catalog.northeastern.edu";
  const concentrationsTag = "#concentrationrequirementstext";
  return $(element)
    .find("li > a")
    .toArray()
    .map((link) => $(link).attr("href"))
    .map((path) => appendPath(base, path, concentrationsTag).toString());
};

/**
 * Converts tables rows into a list of HRows
 * @param $
 * @param table
 */
const tokenizeRows = ($: CheerioStatic, table: CheerioElement): HRow[] => {
  const courseTable: HRow[] = [];

  for (const tr of $(table).find("tbody > tr").toArray()) {
    // different row type
    const type = getRowType($, tr);
    const row = constructRow($, tr, type);
    courseTable.push(row);
  }

  return courseTable;
};

/**
 * Pre-parses the row to determine its type
 * @param $
 * @param tr
 */
const getRowType = ($: CheerioStatic, tr: CheerioElement) => {
  const trClass = tr.attribs["class"];
  const td = $(tr.children[0]);
  const tdClass = td.attr("class");

  if (trClass.includes("subheader")) {
    return HRowType.SUBHEADER;
  } else if (trClass.includes("areaheader")) {
    return HRowType.HEADER;
  } else if (trClass.includes("orclass")) {
    return HRowType.OR_COURSE;
  } else if (tdClass) {
    if (tdClass !== "codecol") {
      throw Error(
        "Expected class to exist with value codecol. Please add this case to the compiler."
      );
    }
    if (td.find(".code").toArray().length > 1) {
      return HRowType.AND_COURSE;
    }
    return HRowType.PLAIN_COURSE;
  }

  const tdText = parseText(td);
  // Different range types
  if (
    RANGE_LOWER_BOUNDED_MAYBE_EXCEPTIONS_1.test(tdText) ||
    RANGE_LOWER_BOUNDED_MAYBE_EXCEPTIONS_2.test(tdText)
  ) {
    return HRowType.RANGE_LOWER_BOUNDED;
  } else if (RANGE_BOUNDED.test(tdText)) {
    return HRowType.RANGE_BOUNDED;
  } else if (RANGE_UNBOUNDED.test(tdText)) {
    return HRowType.RANGE_UNBOUNDED;
  }

  return HRowType.COMMENT;
};

/**
 * Converts a single row based on the passed-in type (determined by {@link getRowType}
 * @param $
 * @param tr
 * @param type
 */
const constructRow = (
  $: CheerioStatic,
  tr: CheerioElement,
  type: HRowType
): HRow => {
  switch (type) {
    case HRowType.HEADER:
    case HRowType.SUBHEADER:
    case HRowType.COMMENT:
      return constructTextRow($, tr, type);
    case HRowType.OR_COURSE:
      return constructOrCourseRow($, tr);
    case HRowType.PLAIN_COURSE:
      return constructPlainCourseRow($, tr);
    case HRowType.AND_COURSE:
      return constructMultiCourseRow($, tr);
    case HRowType.RANGE_LOWER_BOUNDED:
      return constructRangeLowerBoundedMaybeExceptions($, tr);
    case HRowType.RANGE_BOUNDED:
      return constructRangeBounded($, tr);
    case HRowType.RANGE_UNBOUNDED:
      return constructRangeUnbounded($, tr);
    case HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS:
      throw "should never get here";
    default:
      return assertUnreachable(type);
  }
};

const constructTextRow = <T>(
  $: CheerioStatic,
  tr: CheerioElement,
  type: T
): TextRow<T> => {
  const [c1, c2] = ensureLength(2, tr.children).map($);
  const description = parseText(c1);
  const hour = parseHour(c2);
  return { hour, description, type };
};

const constructPlainCourseRow = (
  $: CheerioStatic,
  tr: CheerioElement
): CourseRow<HRowType.PLAIN_COURSE> => {
  const [code, desc, hourCol] = ensureLength(3, tr.children).map($);
  const { subject, classId } = parseCourseTitle(parseText(code));
  const description = parseText(desc);
  const hour = parseHour(hourCol);
  return { hour, description, type: HRowType.PLAIN_COURSE, subject, classId };
};

const constructOrCourseRow = (
  $: CheerioStatic,
  tr: CheerioElement
): CourseRow<HRowType.OR_COURSE> => {
  const [code, desc] = ensureLength(2, tr.children).map($);
  // remove "or "
  const { subject, classId } = parseCourseTitle(
    parseText(code).substring(3).trim()
  );
  const description = parseText(desc);
  // there may be multiple courses in the OR, so we can't backtrack
  return { hour: 0, description, type: HRowType.OR_COURSE, subject, classId };
};

const constructMultiCourseRow = (
  $: CheerioStatic,
  tr: CheerioElement
): MultiCourseRow<HRowType.AND_COURSE> => {
  const [code, desc, hourCol] = ensureLength(3, tr.children).map($);
  const titles = code
    .find(".code")
    .toArray()
    .map($)
    .map(parseText)
    .map(parseCourseTitle);
  const firstDescription = parseText(desc.contents().first());
  const restDescriptions = desc
    .children(".blockindent")
    .toArray()
    // ignore the first four characters, "and "
    .map((c) => parseText($(c)).substring(4).trim());
  const descriptions = [firstDescription, ...restDescriptions];
  if (titles.length !== descriptions.length) {
    const msg = `found titles: ${titles.length} !== found descs: ${descriptions.length}`;
    throw new Error(msg + titles + descriptions);
  }
  const courses = titles.map(({ subject, classId }, i) => ({
    subject,
    classId,
    description: descriptions[i],
  }));
  const hour = parseHour(hourCol);
  return {
    hour,
    type: HRowType.AND_COURSE,
    description: descriptions.join(" and "),
    courses,
  };
};

const constructRangeLowerBoundedMaybeExceptions = (
  $: CheerioStatic,
  tr: CheerioElement
):
  | WithExceptions<
      RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS>
    >
  | RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED> => {
  const [desc, hourCol] = ensureLength(2, tr.children).map($);
  const hour = parseHour(hourCol);
  // text should match one of the following:
  // - CS 9999 or higher[, except CS 9999, CS 9999, CS 3999,... <etc>]
  // - Select from any HIST course numbered 3000 or above.
  // - Complete three HIST courses numbered 2303 or above. Cluster is subject to Department approval.
  const text = parseText(desc);
  // should match the form [["CS 9999", "CS", "9999"], [...]]
  const matches = Array.from(text.matchAll(RANGE_LOWER_BOUNDED_PARSE));
  const [[, subject, , , , id], ...exceptions] = ensureLengthAtLeast(
    1,
    matches
  );
  if (exceptions.length > 0) {
    return {
      type: HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS,
      hour,
      subject,
      classIdStart: Number(id),
      exceptions: exceptions.map(([, subject, , , , id]) => ({
        subject,
        classId: Number(id),
      })),
    };
  }
  return {
    type: HRowType.RANGE_LOWER_BOUNDED,
    hour,
    subject,
    classIdStart: Number(id),
  };
};

const constructRangeBounded = (
  $: CheerioStatic,
  tr: CheerioElement
): RangeBoundedRow<HRowType.RANGE_BOUNDED> => {
  const [desc, hourCol] = ensureLength(2, tr.children).map($);
  const hour = parseHour(hourCol);
  // text should match the form:
  // 1. CS 1000 to CS 5999
  // 2. CS 1000-CS 5999
  const text = parseText(desc);
  // should match the form [["CS 9999", "CS", "9999"], [...]]
  const matches = Array.from(text.matchAll(COURSE_REGEX));
  const [[, subject, classIdStart], [, , classIdEnd]] = ensureLength(
    2,
    matches
  );
  return {
    type: HRowType.RANGE_BOUNDED,
    hour,
    subject,
    classIdStart: Number(classIdStart),
    classIdEnd: Number(classIdEnd),
  };
};

const constructRangeUnbounded = (
  $: CheerioStatic,
  tr: CheerioElement
): RangeUnboundedRow<HRowType.RANGE_UNBOUNDED> => {
  const [desc, hourCol] = ensureLength(2, tr.children).map($);
  const hour = parseHour(hourCol);
  // text should match one of the following:
  // - Any course in ARTD, ARTE, ARTF, ARTG, ARTH, and GAME subject areas as long as prerequisites have been met.
  // - BIOE, CHME, CIVE, EECE, ME, IE, MEIE, and ENGR to Department approval.
  const text = parseText(desc);
  const matches = Array.from(text.match(SUBJECT_REGEX) ?? []);
  const subjects = ensureLengthAtLeast(3, matches);
  return {
    type: HRowType.RANGE_UNBOUNDED,
    hour,
    subjects,
  };
};

const parseHour = (td: Cheerio) => {
  const hourText = td.text();
  return parseInt(hourText.split("-")[0]) || 0;
};
const parseCourseTitle = (parsedCourse: string) => {
  const [subject, classId] = ensureLength(2, parsedCourse.split(" "));
  return {
    subject,
    classId: Number(classId),
  };
};
