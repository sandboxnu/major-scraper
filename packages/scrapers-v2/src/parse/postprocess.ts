import { HRow, HRowType } from "../tokenize/types";
import {
  IAndCourse2,
  IOrCourse2,
  IRequiredCourse,
  Requirement2,
} from "@graduate/common";

// grab the wrapper of the row type enum from tokenize types
type GetRow<RowType> = HRow & { type: RowType };
// sometimes produces a list, if a row token maps to multiple requirements
type Processor<RowType> = (
  tokens: [GetRow<RowType>]
) => Requirement2 | Requirement2[];

// utility for parsing a list of simple courses to IRequiredCourse
const convertCourses = (
  cs: Array<{ subject: string; classId: number }>
): IRequiredCourse[] => {
  return cs.map((c) => ({ ...c, type: "COURSE" }));
};

export const processCourse: Processor<
  HRowType.PLAIN_COURSE | HRowType.OR_COURSE
> = (tokens): IRequiredCourse => {
  const { subject, classId } = tokens[0];
  return { type: "COURSE", classId, subject };
};

export const processRangeLB: Processor<HRowType.RANGE_LOWER_BOUNDED> = (
  tokens
) => {
  const { classIdStart, subject } = tokens[0];
  return {
    type: "RANGE",
    subject,
    idRangeStart: classIdStart,
    idRangeEnd: 9999,
    exceptions: [],
  };
};

export const processRangeLBE: Processor<
  HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS
> = (tokens) => {
  const { subject, classIdStart, exceptions: es } = tokens[0];
  const exceptions = convertCourses(es);
  return {
    type: "RANGE",
    subject,
    idRangeStart: classIdStart,
    idRangeEnd: 9999,
    exceptions,
  };
};

export const processRangeB: Processor<HRowType.RANGE_BOUNDED> = (tokens) => {
  const { subject, classIdStart, classIdEnd } = tokens[0];
  return {
    type: "RANGE",
    subject,
    idRangeStart: classIdStart,
    idRangeEnd: classIdEnd,
    exceptions: [],
  };
};

export const processRangeBE: Processor<
  HRowType.RANGE_BOUNDED_WITH_EXCEPTIONS
> = (tokens) => {
  const { subject, classIdStart, classIdEnd, exceptions: es } = tokens[0];
  const exceptions = convertCourses(es);
  return {
    type: "RANGE",
    subject,
    idRangeStart: classIdStart,
    idRangeEnd: classIdEnd,
    exceptions,
  };
};

export const processRangeU: Processor<HRowType.RANGE_UNBOUNDED> = (tokens) => {
  const { subjects } = tokens[0];
  return subjects.map((subject) => ({
    type: "RANGE",
    subject,
    idRangeStart: 0,
    idRangeEnd: 9999,
    exceptions: [],
  }));
};

export const processOr = (
  tokens: [Requirement2, Requirement2[]]
): IOrCourse2 => {
  const [req, ors] = tokens;
  return { type: "OR", courses: [req, ...ors] };
};

export const processOrOfAnd: Processor<
  HRowType.OR_OF_AND_COURSE | HRowType.AND_COURSE
> = (tokens) => {
  const { courses: cs } = tokens[0];
  const courses = convertCourses(cs);
  return {
    type: "AND",
    courses,
  };
};

export const processAnd = (tokens: [IAndCourse2[]]): IAndCourse2 => {
  const courses = [];
  for (const { courses: cs } of tokens[0]) {
    courses.push(...cs);
  }
  return {
    type: "AND",
    courses,
  };
};

// const Range = (
//   subject: string,
//   idRangeStart: number,
//   idRangeEnd: number,
//   exceptions = []
// ) => {
//   return {
//     exceptions,
//     idRangeEnd,
//     idRangeStart,
//     subject,
//     type: "RANGE",
//   };
// };
// export const processRange = ([range]: [
//   | RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED>
//   | WithExceptions<
//       RangeLowerBoundedRow<HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS>
//     >
//   | RangeBoundedRow<HRowType.RANGE_BOUNDED>
//   | WithExceptions<RangeBoundedRow<HRowType.RANGE_BOUNDED_WITH_EXCEPTIONS>>
//   | RangeUnboundedRow<HRowType.RANGE_UNBOUNDED>
// ]) /*: ICourseRange2*/ => {
//   switch (range.type) {
//     case HRowType.RANGE_UNBOUNDED:
//       break;
//     case HRowType.RANGE_LOWER_BOUNDED:
//       break;
//     case HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS:
//       break;
//     case HRowType.RANGE_BOUNDED:
//       break;
//     case HRowType.RANGE_BOUNDED_WITH_EXCEPTIONS:
//       break;
//   }
// };
