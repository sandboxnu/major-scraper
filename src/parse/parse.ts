import { CompiledRules, Grammar, Parser } from "nearley";
import { HRow } from "../tokenize/types";

// at runtime, generate the ./grammar.ts file from the grammar.ne file
// eslint-disable-next-line @typescript-eslint/no-var-requires
const grammar: CompiledRules = require("./grammar");

export const parseRows = (rows: HRow[]) => {
  const parser = new Parser(Grammar.fromCompiled(grammar));

  // according to docs, "you would feed a Parser instance an array of objects"
  // https://nearley.js.org/docs/tokenizers#custom-token-matchers
  // however signature only takes string, so cast to any
  parser.feed(rows as any);

  // make sure there are no multiple solutions, as our grammar should be unambiguous
  if (parser.results.length === 0) {
    throw new Error("unexpected end of tokens");
  } else if (parser.results.length === 1) {
    return parser.results[0];
  }
  throw new Error(`${parser.results.length} solutions, grammar is ambiguous`);
};
