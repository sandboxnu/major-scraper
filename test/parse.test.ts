import { describe, expect, test } from "vitest";
import { parseRows, parseTokens } from "@/parse";
import { type HRow, type HSection } from "@/tokenize";
import {
  AND_COURSE_PARSED,
  AND_COURSE_TOKEN,
  HEADER_TOKEN,
  OR_COURSE_PARSED,
  OR_COURSE_TOKEN,
  OR_OF_AND_COURSE_PARSED,
  OR_OF_AND_COURSE_TOKEN,
  PLAIN_COURSE_PARSED,
  PLAIN_COURSE_TOKEN,
  RANGE_BOUNDED_PARSED,
  RANGE_BOUNDED_TOKEN,
  RANGE_BOUNDED_WITH_EXCEPTIONS_PARSED,
  RANGE_BOUNDED_WITH_EXCEPTIONS_TOKEN,
  RANGE_LOWER_BOUNDED_PARSED,
  RANGE_LOWER_BOUNDED_TOKEN,
  RANGE_LOWER_BOUNDED_WITH_EXCEPTION_PARSED,
  RANGE_LOWER_BOUNDED_WITH_EXCEPTION_TOKEN,
  RANGE_UNBOUNDED_PARSED,
  RANGE_UNBOUNDED_TOKEN,
  SUBHEADER_TOKEN,
  X_OF_MANY_TOKEN,
} from "test/parseConstant";
import type { Requirement2 } from "@/types";
import {
  ARCH_ENGLISH_2022,
  BIOENG_BIOCHEM_2022,
  CHEMICAL_ENG_2022,
  CS_BSCS_2022,
  CS_BUSINESS_ADMIN_2022,
  CS_ENGLISH_2023,
  CS_HISTORY_2022,
  CS_MATH_2022,
  CS_MUSIC_2023,
  HUMAN_SERVICE_INTER_AFFAIRS_2023,
  MEDIA_SCREEN_STUDIES_HISTORY_2022,
  PHYSICS_2022,
  formatParsePath,
  formatTokenPath,
} from "test/testUrls";
import { readFile } from "fs/promises";

function expectReqs(testName: string, tokens: HRow[], reqs: Requirement2[]) {
  test(testName, () => {
    expect(parseRows("", [HEADER_TOKEN, ...tokens])).toStrictEqual<
      Requirement2[]
    >([
      {
        type: "SECTION",
        title: "Test Requirement",
        requirements: reqs,
        minRequirementCount: reqs.length,
      },
    ]);
  });
}

function testParse(testName: string, tokens: HRow[], reqs: Requirement2[]) {
  test(testName, () => {
    expect(parseRows("", tokens)).toStrictEqual<Requirement2[]>(reqs);
  });
}

async function testParseSnapshot(testName: string, fileName: string) {
  const json = JSON.parse(
    await readFile(`./test/tokenized/${fileName}.json`, "utf-8"),
  ) as { sections: HSection[] };

  test(testName, async () => {
    await expect(
      JSON.stringify(parseTokens(json.sections), null, 2),
    ).toMatchFileSnapshot(formatParsePath(fileName));
  });
}

describe("Parse Phase: Unit tests", () => {
  expectReqs(
    "course",
    [PLAIN_COURSE_TOKEN, PLAIN_COURSE_TOKEN],
    [PLAIN_COURSE_PARSED, PLAIN_COURSE_PARSED],
  );

  describe("orCourse", () => {
    expectReqs(
      "single OR course",
      [PLAIN_COURSE_TOKEN, OR_COURSE_TOKEN],
      [
        {
          type: "OR",
          courses: [PLAIN_COURSE_PARSED, OR_COURSE_PARSED],
        },
      ],
    );

    expectReqs(
      "aggregate multiple consecutive OR course into one requirement",
      [PLAIN_COURSE_TOKEN, OR_COURSE_TOKEN, OR_COURSE_TOKEN, OR_COURSE_TOKEN],
      [
        {
          type: "OR",
          courses: [
            PLAIN_COURSE_PARSED,
            OR_COURSE_PARSED,
            OR_COURSE_PARSED,
            OR_COURSE_PARSED,
          ],
        },
      ],
    );

    expectReqs(
      "AND course followed by an OR course",
      [AND_COURSE_TOKEN, OR_COURSE_TOKEN],
      [
        {
          type: "OR",
          courses: [AND_COURSE_PARSED, OR_COURSE_PARSED],
        },
      ],
    );

    expectReqs(
      "AND course followed by an OR OF AND course",
      [AND_COURSE_TOKEN, OR_OF_AND_COURSE_TOKEN],
      [
        {
          type: "OR",
          courses: [AND_COURSE_PARSED, OR_OF_AND_COURSE_PARSED],
        },
      ],
    );
  });

  describe("andCourse", () => {
    expectReqs(
      "single AND course",
      [AND_COURSE_TOKEN, PLAIN_COURSE_TOKEN],
      [AND_COURSE_PARSED, PLAIN_COURSE_PARSED],
    );

    expectReqs(
      "multiple AND courses",
      [AND_COURSE_TOKEN, AND_COURSE_TOKEN],
      [AND_COURSE_PARSED, AND_COURSE_PARSED],
    );
  });

  describe("rangeCourse", () => {
    expectReqs("unbounded", [RANGE_UNBOUNDED_TOKEN], RANGE_UNBOUNDED_PARSED);

    expectReqs(
      "lower bounded",
      [RANGE_LOWER_BOUNDED_TOKEN],
      [RANGE_LOWER_BOUNDED_PARSED],
    );

    expectReqs(
      "lower bounded with exceptions",
      [RANGE_LOWER_BOUNDED_WITH_EXCEPTION_TOKEN],
      [RANGE_LOWER_BOUNDED_WITH_EXCEPTION_PARSED],
    );

    expectReqs("bounded", [RANGE_BOUNDED_TOKEN], [RANGE_BOUNDED_PARSED]);

    expectReqs(
      "bounded with exceptions",
      [RANGE_BOUNDED_WITH_EXCEPTIONS_TOKEN],
      [RANGE_BOUNDED_WITH_EXCEPTIONS_PARSED],
    );
  });

  describe("X of Many", () => {
    expectReqs(
      "no token before XOM token",
      [
        X_OF_MANY_TOKEN,
        PLAIN_COURSE_TOKEN,
        AND_COURSE_TOKEN,
        RANGE_BOUNDED_TOKEN,
        RANGE_BOUNDED_WITH_EXCEPTIONS_TOKEN,
      ],
      [
        {
          type: "XOM",
          numCreditsMin: 8,
          courses: [
            PLAIN_COURSE_PARSED,
            AND_COURSE_PARSED,
            RANGE_BOUNDED_PARSED,
            RANGE_BOUNDED_WITH_EXCEPTIONS_PARSED,
          ],
        },
      ],
    );

    expectReqs(
      "some token before XOM token",
      [
        PLAIN_COURSE_TOKEN,
        AND_COURSE_TOKEN,
        X_OF_MANY_TOKEN,
        PLAIN_COURSE_TOKEN,
        AND_COURSE_TOKEN,
        RANGE_BOUNDED_TOKEN,
        RANGE_BOUNDED_WITH_EXCEPTIONS_TOKEN,
      ],
      [
        PLAIN_COURSE_PARSED,
        AND_COURSE_PARSED,
        {
          type: "XOM",
          numCreditsMin: 8,
          courses: [
            PLAIN_COURSE_PARSED,
            AND_COURSE_PARSED,
            RANGE_BOUNDED_PARSED,
            RANGE_BOUNDED_WITH_EXCEPTIONS_PARSED,
          ],
        },
      ],
    );
  });

  describe("subsections", () => {
    testParse(
      "no section info and courses before subsection",
      [
        HEADER_TOKEN,
        SUBHEADER_TOKEN,
        AND_COURSE_TOKEN,
        PLAIN_COURSE_TOKEN,
        SUBHEADER_TOKEN,
        PLAIN_COURSE_TOKEN,
        RANGE_BOUNDED_TOKEN,
      ],
      [
        {
          type: "SECTION",
          title: "Test Requirement",
          requirements: [
            {
              type: "SECTION",
              title: "Subheader Token",
              requirements: [AND_COURSE_PARSED, PLAIN_COURSE_PARSED],
              minRequirementCount: 2,
            },
            {
              type: "SECTION",
              title: "Subheader Token",
              requirements: [PLAIN_COURSE_PARSED, RANGE_BOUNDED_PARSED],
              minRequirementCount: 2,
            },
          ],
          minRequirementCount: 2,
        },
      ],
    );

    testParse(
      "courses before subsections",
      [
        HEADER_TOKEN,
        PLAIN_COURSE_TOKEN,
        AND_COURSE_TOKEN,
        RANGE_BOUNDED_WITH_EXCEPTIONS_TOKEN,
        SUBHEADER_TOKEN,
        AND_COURSE_TOKEN,
        PLAIN_COURSE_TOKEN,
        SUBHEADER_TOKEN,
        PLAIN_COURSE_TOKEN,
        RANGE_BOUNDED_TOKEN,
      ],
      [
        {
          type: "SECTION",
          title: "Test Requirement",
          requirements: [
            PLAIN_COURSE_PARSED,
            AND_COURSE_PARSED,
            RANGE_BOUNDED_WITH_EXCEPTIONS_PARSED,
            {
              type: "SECTION",
              title: "Subheader Token",
              requirements: [AND_COURSE_PARSED, PLAIN_COURSE_PARSED],
              minRequirementCount: 2,
            },
            {
              type: "SECTION",
              title: "Subheader Token",
              requirements: [PLAIN_COURSE_PARSED, RANGE_BOUNDED_PARSED],
              minRequirementCount: 2,
            },
          ],
          minRequirementCount: 5,
        },
      ],
    );
  });
});

describe("Parse Phase: 2022 Snapshots", async () => {
  await testParseSnapshot(
    "Computer Science BSCS (science requirements)",
    CS_BSCS_2022,
  );

  await testParseSnapshot(
    "CS & Business (nested linked concentration page)",
    CS_BUSINESS_ADMIN_2022,
  );
  await testParseSnapshot("Physics (3 courses per AND)", PHYSICS_2022);
  await testParseSnapshot(
    "Media screen & study history (range bounded)",
    MEDIA_SCREEN_STUDIES_HISTORY_2022,
  );
  await testParseSnapshot(
    "CS and Math (range bounded with exceptions)",
    CS_MATH_2022,
  );
  await testParseSnapshot(
    "CS and History (range lower bounded)",
    CS_HISTORY_2022,
  );
  await testParseSnapshot(
    "Chemical Engineering (range unbounded)",
    CHEMICAL_ENG_2022,
  );
  await testParseSnapshot("Bioeng Biochem (Or of ANDs)", BIOENG_BIOCHEM_2022);
  await testParseSnapshot(
    "Architecture and English (no tabs)",
    ARCH_ENGLISH_2022,
  );
});

describe("Parse Phase: 2023 Snapshots", async () => {
  await testParseSnapshot(
    "CS and Music (section info phrasing in middle of a section)",
    CS_MUSIC_2023,
  );

  await testParseSnapshot(
    "CS and English (subsubsection aka subsection with indentation)",
    CS_ENGLISH_2023,
  );

  await testParseSnapshot(
    "Human Service and International Affairs (separated section info and course tables)",
    HUMAN_SERVICE_INTER_AFFAIRS_2023,
  );
});
