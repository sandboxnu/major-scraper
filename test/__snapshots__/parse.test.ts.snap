// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse and 1`] = `
Array [
  Object {
    "courses": Array [
      Object {
        "classId": 2500,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2501,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2510,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2511,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 3500,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2501,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
    ],
    "type": "AND",
  },
]
`;

exports[`parse and and or 1`] = `
Array [
  Object {
    "courses": Array [
      Object {
        "classId": 2500,
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2501,
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2501,
        "subject": "CS",
        "type": "COURSE",
      },
    ],
    "type": "OR",
  },
  Object {
    "courses": Array [
      Object {
        "classId": 2500,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2501,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2510,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2511,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 3500,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
      Object {
        "classId": 2501,
        "description": "",
        "subject": "CS",
        "type": "COURSE",
      },
    ],
    "type": "AND",
  },
]
`;
